### PR TITLE
Delete twap history

### PR DESCRIPTION
**Pull request by mattverse**
This PR Adds removing old twap history that would be saved in the state machine.

The following are set as Params of the gamm module for twap history.

_Spec for Gamm module to be added._

- The number of twap pool it tends go over and delete per iteration
- Twap history Deletion Interval(Per how many blocks would it be activating deletion iteration)
- How long would twap history be kept

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -486,6 +488,7 @@ func NewOsmosisApp(
 		lockuptypes.ModuleName,
 		crisistypes.ModuleName, govtypes.ModuleName, stakingtypes.ModuleName, claimtypes.ModuleName,
 		// Note: epochs' endblock should be "real" end of epochs, we keep epochs endblock at the end
+		gammtypes.ModuleName,
```
Comment (ValarDragon) [inline app/app.go:491]: Can you move the epochs comment below the gamm types line

Comment (mattverse) [reply app/app.go:491]: srry my bad!

File: x/gamm/genesis.go [hunk 2]
```diff
@@ -14,6 +14,7 @@ import (
 func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState, unpacker codectypes.AnyUnpacker) {
 	k.SetParams(ctx, genState.Params)
 	k.SetNextPoolNumber(ctx, genState.NextPoolNumber)
+	k.SetNextTwapHistoryDeleteIndex(ctx, genState.NextTwapHistoryDeletePoolId)
```
Comment (ValarDragon) [inline x/gamm/genesis.go:17]: Should set it to `0` or smth if genstate is blank? I have no idea what to set it to atm

File: x/gamm/keeper/pool_twap.go [hunk 3]
```diff
@@ -182,6 +183,7 @@ func (k Keeper) GetRecentPoolTwapSpotPrice(
 		return sdk.Dec{}, fmt.Errorf("pool twap history prior to current time does not exist")
 	}
 
+	// TODO: use duration.Seconds?
 	desiredTime := ctx.BlockTime().Add(-duration * time.Second)
 	desiredTimeAdjacentPoolTwap, exists := k.GetPoolTwapHistory(ctx, poolId, desiredTime.Add(time.Second))
```
Comment (ValarDragon) [inline x/gamm/keeper/pool_twap.go:188]: Can you add some comments explaining desired time and why this mysterious time.second is added?

Should it be time.nanosecond or smth added?

File: x/gamm/keeper/pool_twap_test.go [hunk 4]
```diff
@@ -494,10 +494,135 @@ func (suite *KeeperTestSuite) TestPoolTwapSpotPrice() {
...
+				_, _, err = suite.app.GAMMKeeper.SwapExactAmountIn(ctx, acc1, pool3, barcoin, "baz", sdk.ZeroInt())
+				suite.Require().NoError(err)
+
+				ctx = suite.ctx.WithBlockTime(time.Now().Add(time.Hour*1 + time.Minute*40))
+				_, _, err = suite.app.GAMMKeeper.SwapExactAmountIn(ctx, acc1, pool1, foocoin, "bar", sdk.ZeroInt())
+				suite.Require().NoError(err)
+
+				ctx = suite.ctx.WithBlockTime(time.Now().Add(time.Hour*1 + time.Minute*50))
+				_, _, err = suite.app.GAMMKeeper.SwapExactAmountIn(ctx, acc1, pool2, foocoin, "bar", sdk.ZeroInt())
+				suite.Require().NoError(err)
```
Comment (ValarDragon) [inline x/gamm/keeper/pool_twap_test.go:579]: Can we put this into a loop thats like `num_twap_entries = 12`, `for i := 0; i < num_twap_entries; i++ { {do thing with blocktime being time.Minute * (10 * num_twap_entries) } }`

File: x/gamm/keeper/pool_twap.go [hunk 5]
```diff
@@ -244,7 +246,78 @@ func (k Keeper) GetRecentPoolTwapSpotPrice(
...
+		if err != nil {
+			panic(err)
+		}
+		nextTwapHistoryIndex = val.GetValue()
+	}
+
+	return nextTwapHistoryIndex
+}
+
+// DeleteTwapHisotry deletes any twap history records that are beyond currentTime + keepDuration
```
Comment (ValarDragon) [inline x/gamm/keeper/pool_twap.go:278]:
```suggestion
// DeleteTwapHistory deletes any twap history records that are beyond currentTime + keepDuration
```

File: x/gamm/genesis.go (hunk 2 above)
Comment (mattverse) [reply x/gamm/genesis.go:17]: Do you mean the default genesis state? The default genesis state is 
```
func DefaultGenesis() *GenesisState {
	return &GenesisState{
		Pools:                       []*codectypes.Any{},
		NextPoolNumber:              1,
		Params:                      DefaultParams(),
		NextTwapHistoryDeletePoolId: 1,
	}
}
```

File: x/gamm/keeper/pool_twap.go (hunk 3 above)
Comment (mattverse) [reply x/gamm/keeper/pool_twap.go:188]: Sure! Definitely should make this clearer

File: x/gamm/keeper/pool_twap_test.go (hunk 4 above)
Comment (mattverse) [reply x/gamm/keeper/pool_twap_test.go:579]: Or do you think I should make the test case itself more simpler?

File: x/gamm/genesis.go (hunk 2 above)
Comment (ValarDragon) [reply x/gamm/genesis.go:17]: Oh cool

File: x/gamm/simulation/operations.go [hunk 6]
```diff
@@ -157,7 +157,10 @@ func SimulateMsgCreatePool(ak stakingTypes.AccountKeeper, bk stakingTypes.BankKe
 
 		// set the pool params to set the pool creation fee to dust amount of denom
 		k.SetParams(ctx, types.Params{
-			PoolCreationFee: sdk.Coins{sdk.NewInt64Coin(denoms[0], 1)},
+			PoolCreationFee:             sdk.Coins{sdk.NewInt64Coin(denoms[0], 1)},
+			NumTwapHistoryPerDeletion:   10,
+			TwapHistoryDeletionInterval: 5,
+			TwapHistoryKeepDuration:     time.Duration(time.Hour * 24),
```
Comment (mattverse) [inline x/gamm/simulation/operations.go:163]: I don't think these all need to go in Params, but we do need further discussion about what default value they should have

File: x/gamm/keeper/pool_twap.go [hunk 7]
```diff
@@ -244,7 +246,78 @@ func (k Keeper) GetRecentPoolTwapSpotPrice(
 	return twap, nil
 }
 
-// function for validating whether specic string(s) are contained in a slice
+// SetNextTwapHistoryDeleteIndex sets the next twa history for deletion
```
Comment (mattverse) [inline x/gamm/keeper/pool_twap.go:249]: // SetNextTwapHistoryDeleteIndex sets the next twap history for deletion

**Comment by daniel-farina** [comment]
We will re-plan the twap architecture.

**Comment by mattverse** [comment]
Should address https://github.com/osmosis-labs/osmosis/tree/twap as well and discuss how we can improve things

**Pull request closed without merging**